Fix redecl chain of classes and class templates

martong opened the pull request:
The crux of the issue was that lookup could not find previous decls of
a friend class. The solution involves making the friend declarations
visible in their decl context (i.e. adding them to the lookup table).
This fix involves two other repairs:
(1) We could not handle the addition of injected class types properly
when a redecl chain was involved, now this is fixed.
(2) DeclContext::removeDecl failed if the lookup table in Vector form
did not contain the to be removed element. This caused troubles in
ASTImporter::ImportDeclContext. This is also fixed.

Fixes #484

martong commented:
Unfortunately dos not fix #487 . Perhaps because there are other cases (other than broken redecl chains of classes) when structural equivalency does not match and thus an empty DeclGroup is created.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2431,7 +2449,7 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
     } else {
       if (GetImportedOrCreateDecl(D2, D, Importer.getToContext(),
                                   D->getTagKind(), DC, StartLoc, Loc,
-                                  Name.getAsIdentifierInfo(), PrevDecl))
```
This is a mistake here.

martong commented on lib/Analysis/CFG.cpp:
```diff
@@ -2248,6 +2248,11 @@ CFGBlock *CFGBuilder::VisitConditionalOperator(AbstractConditionalOperator *C,
 }
 
 CFGBlock *CFGBuilder::VisitDeclStmt(DeclStmt *DS) {
+  //if (DS->isSingleDecl() && !DS->getSingleDecl())
```
Should be part of another PR.

martong commented on lib/Basic/SourceManager.cpp:
```diff
@@ -2061,6 +2061,7 @@ bool SourceManager::isBeforeInTranslationUnit(SourceLocation LHS,
       return LIsScratch;
     return LOffs.second < ROffs.second;
   }
+  return LB < RB;
```
Leftover from debug :(

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4120,6 +4184,100 @@ TEST_P(ImportVariables, InitAndDefinitionAreInTheFromContext) {
+  EXPECT_EQ(Imported0, Imported1->getPreviousDecl());
+}
+
+// TODO test class template specializations too
```
Leftover, sorry :(

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2406,9 +2415,18 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-          Importer.getToContext().getInjectedClassNameType(D2CXX,
-              ToDescribed->getInjectedClassNameSpecialization());
+          // Create an injected type for the whole redecl chain.
+          SmallVector<Decl*, 2> Redecls = getCanonicalForwardRedeclChain(D2CXX);
```
SmallVector<Decl *, 2>

balazske commented on include/clang/AST/DeclContextInternals.h:
```diff
@@ -122,6 +122,13 @@ struct StoredDeclsList {
              == Vec.end() && "list still contains decl");
   }
 
+  bool containsInVector(NamedDecl* D) {
```
`NamedDecl *D`

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4549,18 +4569,24 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
+  if (FoundTemplate) {
+    auto *Recent =
+        const_cast<ClassTemplateDecl *>(FoundTemplate->getMostRecentDecl());
+    D2->setPreviousDecl(Recent);
```
I think the decl chains of template and templated Decls should match. Probably add assert here?
`Recent->getTemplatedDecl() == ToTemplated->getPreviousDecl()`

martong replied:
This assert was a very good idea, it revealed a mistake I made, working on fixing that.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4480,17 +4498,19 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
 
       Decl *Found = FoundDecls[I];
-      if (auto *FoundTemplate = dyn_cast<ClassTemplateDecl>(Found)) {
+      if ((FoundTemplate = dyn_cast<ClassTemplateDecl>(Found))) {
```
Remove one level of unneeded `(` `)`.

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4032,6 +4062,40 @@ TEST_P(DeclContextTest, removeDeclOfClassTemplateSpecialization) {
+  auto *Map = DC->getLookupPtr();
+  ASSERT_TRUE(Map);
+  auto I = Map->find(A0->getDeclName());
+  ASSERT_FALSE(I == Map->end());
```
`ASSERT_EQ` ?

martong commented:
Update: Changed the style where you indicated, and added an assert, in a new commit so you can follow the diff.

balazske commented:
The code looks OK.
But in the tests it could be better to use descriptive names (not `Xxx0` and `Xxx1`) and not re-use existing variables with different meaning. (If not corrected here, this can be problem in the opensource review. My test code was not accepted with such problems once.)

balazske commented:
Another thing: We can check if there are more similar tests that are not in the `ImportFriendClasses` fixture and move these in. (Similar issues may exist with other tests.)

Commit:
Fix redecl chain of classes and class templates

The crux of the issue was that lookup could not find previous decls of
a friend class. The solution involves making the friend declarations
visible in their decl context (i.e. adding them to the lookup table).
This fix involves two other repairs:
(1) We could not handle the addition of injected class types properly
when a redecl chain was involved, now this is fixed.
(2) DeclContext::removeDecl failed if the lookup table in Vector form
did not contain the to be removed element. This caused troubles in
ASTImporter::ImportDeclContext. This is also fixed.

martong commented:
Update: changed the PR to properly build up the redecl chain not just in case of Friends.
Now this seems to fix #487 too.
I am going to create additional tests as well.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2281,35 +2282,30 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-                // FoundDef may not have every implicit method that D has
-                // because implicit methods are created only if they are used.
-                ImportImplicitMethods(DCXX, FoundCXX);
+          if (D->isThisDeclarationADefinition()) {
```
`isThisDeclarationADefinition` and `isCompleteDefinition` return the same value. Something can be simplified here?

martong replied:
Good point, thanks.

martong replied:
This part is really messy so would be very good if we could simplify somehow.

Commit:
Rename variables in test

Commit:
Regroup test cases to ImportFriendClasses

Commit:
Simplify VisitRecordDecl

Commit:
Use similar pattern which had been used in VisitFunctionDecl

Commit:
Add tests for anon and unnamed structs

Commit:
Fix lldb TestPersistentTypes.py

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2282,52 +2282,28 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-              !FoundRecord->isCompleteDefinition())
-            FoundRecord->getASTContext().getExternalSource()->CompleteType(FoundRecord);
-          if (D->hasExternalLexicalStorage())
-            D->getASTContext().getExternalSource()->CompleteType(D);
```
Are the `CompleteType` calls not needed?

martong replied:
We already completed `D` in line 2255.
And `FoundRecord` must had been completed either during it's import or it is created by the parser and don't have external lexical storage.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2273,67 +2274,43 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
+        //    struct { struct A *next; } entry0;
+        //    struct { struct A *next; } entry1;
+        //  };
+        //  struct X { struct { int a; }; struct { int b; }; // anon structs
```
A `}` (close of `X`) is missing from comment above.

martong replied:
Thanks, good point.

Commit:
Fix lldb TestSharedLib.py

Commit:
Fix comment missing '}'

Commit:
Add tests for class and class template chains

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4087,6 +4087,305 @@ TEST_P(ImportVariables, InitAndDefinitionAreInTheFromContext) {
+  EXPECT_FALSE(ToProto->isThisDeclarationADefinition());
+  EXPECT_EQ(ToDef->getPreviousDecl(), ToProto);
+}
+
```
Good tests.

Commit:
Add comment

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4087,6 +4087,305 @@ TEST_P(ImportVariables, InitAndDefinitionAreInTheFromContext) {
+  auto ToD = LastDeclMatcher<ClassTemplateDecl>().match(ToTU, Pattern);
+  EXPECT_TRUE(ImportedD == ToD);
+  EXPECT_FALSE(ToD->isThisDeclarationADefinition());
+  EXPECT_TRUE(ToD->getTemplatedDecl());
```
This check is not needed or is needed as the first check. It is not "normal" that there is no `getTemplatedDecl` (`isThisDeclarationADefinition` calls also getTemplatedDecl).

martong replied:
Ok, good point, I moved them upwards and made the check to be an `ASSERT_`.

Commit:
Move getTemplatedDecl() calls upwards

Commit:
Change one EXPECT to ASSERT

martong commented:
ctu-clang7:
 https://github.com/Ericsson/clang/commit/9d42ac3de0e1292ad018d12fd4ef57d803e42f25

martong commented:
https://reviews.llvm.org/D53655
